src/dvc_data/index/diff.py

```diff
@@ -255,35 +256,39 @@ def _detect_renames(changes: Iterable[Change]):
     def _get_key(change):
         return change.key
 
-    added[:] = sorted(added, key=_get_key)
-    deleted[:] = sorted(deleted, key=_get_key)
+    # Sort the lists to maintain the same order
+    # as older implementation.
+    added.sort(key=_get_key)
+    deleted.sort(key=_get_key)
 
-    for change in added:
-        new_entry = change.new
-        assert new_entry
+    # Create a dictionary for fast lookup of deletions by hash_info
+    deleted_dict: dict[Optional[HashInfo], deque[Change]] = defaultdict(deque)
+    for change in deleted:
+        change_hash = change.old.hash_info if change.old else None
+        # appendleft to get queue behaviour (we pop off right)
+        deleted_dict[change_hash].appendleft(change)
 
-        if not new_entry.hash_info:
-            yield change
-            continue
+    for change in added:
+        new_hash_info = change.new.hash_info if change.new else None
 
-        index, old_entry = None, None
-        for idx, ch in enumerate(deleted):
-            assert ch.old
-            if ch.old.hash_info == new_entry.hash_info:
-                index, old_entry = idx, ch.old
-                break
+        # If the new entry is the same as a deleted change,
+        # it is in fact a rename.
+        # Note: get instead of __getitem__, to avoid creating
+        # unnecessary entries.
```

Comment:
What do you mean by this?

Reply:
Looks like this comment got attached to different line for some reason. I am talking about `Note:`.

```
# Note: get instead of __getitem__, to avoid creating
# unnecessary entries.
```

Reply:
Oh, okay. I did not notice `defaultdict` earlier. Ok with the comment then.